api/pkg/apis/v1alpha1/managers/campaigns/campaigns-manager.go

```diff
@@ -96,6 +99,65 @@ func (m *CampaignsManager) UpsertState(ctx context.Context, name string, state m
 	}
 	state.ObjectMeta.FixNames(name)
 
+	metadata := map[string]interface{}{
+		"namespace": state.ObjectMeta.Namespace,
+		"group":     model.WorkflowGroup,
+		"version":   "v1",
+		"resource":  "campaigns",
+		"kind":      "Campaign",
+	}
+
+	if state.Spec != nil {
+		rootResource := state.Spec.RootResource
+		if rootResource != "" {
+			log.Debugf(" M (Campaigns): campaign root resource: %s, campaign: %s", rootResource, name)
+			resourceName := "campaigncontainers"
+			kind := "CampaignContainer"
+			containerMetadata := map[string]interface{}{
+				"version":   "v1",
+				"group":     model.WorkflowGroup,
+				"resource":  resourceName,
+				"namespace": state.ObjectMeta.Namespace,
+				"kind":      kind,
+			}
+			getRequest := states.GetRequest{
+				ID:       rootResource,
+				Metadata: containerMetadata,
+			}
+			_, err = m.StateProvider.Get(ctx, getRequest)
+			if err != nil {
+				log.Debugf(" M (Campaigns): get campaign container %s, err %v", rootResource, err)
+			}
+			cErr, ok := err.(v1alpha2.COAError)
+			if ok && cErr.State == v1alpha2.NotFound {
```

Comment:
So, we automatically create a container if the container doesn't exist? Why are we doing this?